File: 559_MaximumDepthOfN-aryTree/main.cpp
/*
// Definition for a Node
class Node{
public:
    int val;
    vector<Node*> children;
    
    Node(){}
    
    Node(int _val, vector<Node*> _children){
        val = _val;
        children = _children;
    }
};

class Solution{
public:
    int maxDepth(Node* root){
        int maxDepth = 0;
        maxDepthHelper(root, 0, maxDepth);
        return maxDepth
    }
    
private:
    void maxDepthHelper(Node* root, int depth, int& maxDepth){
        if(root){
            depth++;
            if(root->children.empty()) maxDepth = max(maxDepth, depth);
            for(const auto c : root->children) maxDepthHelper(c, depth, maxDepth);
        }
    }
};

File: 559_MaximumDepthOfN-aryTree/main.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution{
    private int maxDepth = 0;
    
    public int maxDepth(Node root){
        maxDepthHelper(root, 0);
        return maxDepth;
    }

    private void maxDepthHelper(Node root, int depth){
        if(root != null){
            depth++;
            if(root.children.isEmpty()) maxDepth = Math.max(maxDepth, depth);
            for(Node c : root.children) maxDepthHelper(c, depth);
        }
    }
}
